include/flang/lower/ASTBuilder.h

```diff
@@ -0,0 +1,332 @@
+  /// Set that the construct `cstr` (if not a nullptr) has branches.
+  static void setBranches(Evaluation *cstr) {
+    if (cstr) {
+      cstr->setBranches();
+    }
+  }
+
+  EvalVariant u;
+  ParentType parent;
+  parser::CharBlock pos;
+  std::optional<parser::Label> lab;
+  std::list<Evaluation> *subs{nullptr}; // construct sub-statements
+  CFGAnnotation cfg{CFGAnnotation::None};
+  bool isTarget{false};         // this evaluation is a control target
+  bool containsBranches{false}; // construct contains branches
+};
+
+/// A program is a list of program units.
+/// These units can be function like, module like, or block data
+struct ProgramUnit {
+  template <typename A>
+  ProgramUnit(A *ptr, const ParentType &parent) : p{ptr}, parent{parent} {}
+
+  std::variant<const parser::MainProgram *, const parser::FunctionSubprogram *,
+               const parser::SubroutineSubprogram *, const parser::Module *,
+               const parser::Submodule *,
+               const parser::SeparateModuleSubprogram *,
+               const parser::BlockData *>
+      p;
+  ParentType parent;
+};
+
+/// Function-like units have similar structure. They all can contain executable
+/// statements.
+struct FunctionLikeUnit : public ProgramUnit {
+  // wrapper statements for function-like syntactic structures
+  using FunctionStatement =
+      std::variant<const parser::Statement<parser::ProgramStmt> *,
+                   const parser::Statement<parser::EndProgramStmt> *,
+                   const parser::Statement<parser::FunctionStmt> *,
+                   const parser::Statement<parser::EndFunctionStmt> *,
+                   const parser::Statement<parser::SubroutineStmt> *,
+                   const parser::Statement<parser::EndSubroutineStmt> *,
+                   const parser::Statement<parser::MpSubprogramStmt> *,
+                   const parser::Statement<parser::EndMpSubprogramStmt> *>;
+
+  FunctionLikeUnit(const parser::MainProgram &f, const ParentType &parent);
+  FunctionLikeUnit(const parser::FunctionSubprogram &f,
+                   const ParentType &parent);
+  FunctionLikeUnit(const parser::SubroutineSubprogram &f,
+                   const ParentType &parent);
+  FunctionLikeUnit(const parser::SeparateModuleSubprogram &f,
+                   const ParentType &parent);
+
+  bool isMainProgram() {
+    return std::get_if<const parser::Statement<parser::EndProgramStmt> *>(
+        &funStmts.back());
+  }
+  const parser::FunctionStmt *isFunction() {
+    return isA<parser::FunctionStmt>();
+  }
+  const parser::SubroutineStmt *isSubroutine() {
+    return isA<parser::SubroutineStmt>();
+  }
+  const parser::MpSubprogramStmt *isMPSubp() {
+    return isA<parser::MpSubprogramStmt>();
+  }
+
+  const semantics::Scope *scope{nullptr}; // scope from front-end
+  std::list<FunctionStatement> funStmts;  // begin/end pair
+  std::list<Evaluation> evals;            // statements
+  std::list<FunctionLikeUnit> funcs;      // internal procedures
```

Comment:
See previous comment about `std::list`. If the reason for using `std::list` is that these need push and pop from both the back and front then `std::deque` may be a better choice. However in this PR I only see peek at the front and back and push at the back, in which case `std::vector` is probably more appropriate.

Reply:
By the way, the parser uses `std::list`, so the project as a whole hasn't shied away from using that class.

This code is work-in-progress. This PR should not be taken as a complete enumeration of the requirements as to what this data structure may eventually constitute or how it will ultimately be used. (It's already changed, fwiw.) There actually was code that did insertions of nodes into the `list` structure at one point.

We'll make a note that some performance tuning might be in order here.

Reply:
Regarding the `std::list` vs `std::vector`, there is another a reason for actually using a list right now; the address of elements inside the lists are taken.  See:

- `FunctionLikeUnit::evals` elements address are taken here: https://github.com/jeanPerier/f18/blob/10e01a4e505fe6f874dffe4d5f08d2503a695576/lib/lower/ASTBuilder.cpp#L308

- `ModuleLikeUnit::funcs` elements address are taken here: https://github.com/jeanPerier/f18/blob/10e01a4e505fe6f874dffe4d5f08d2503a695576/lib/lower/ASTBuilder.cpp#L323

- `FunctionLikeUnit::funcs` elements are indirectly taken (ASTBuilder::funclist points to funcs, address of element of *funclist are taken) at:
https://github.com/jeanPerier/f18/blob/10e01a4e505fe6f874dffe4d5f08d2503a695576/lib/lower/ASTBuilder.cpp#L291

- `Program::units` elements address here: 
https://github.com/jeanPerier/f18/blob/10e01a4e505fe6f874dffe4d5f08d2503a695576/lib/lower/ASTBuilder.cpp#L323

Because `std::vector` element addresses may be [invalidated after insertion](https://en.cppreference.com/w/cpp/container#Iterator_invalidation), that would be a risky choice here.

Reply:
If the addresses are taken and we might resize and invalidate those addresses, `std::deque` can be used instead of `std::list` as that has the same guarantee that addresses will not be invalidated after resizing.

Reply:
`std::list` was used intentionally. The reason for that was to allow O(1) insertions anywhere in the structure to simplify lowering.